weather_aplication/SearchViewController.swift

```diff
@@ -0,0 +1,152 @@
+import SnapKit
+import UIKit
+
+struct Location {
+    public var city: String?
+}
+
+class SearchViewController: UIViewController {
+    private let kSideMargin: CGFloat = 10
+    private let kTopMargin: CGFloat = 20
+    private let kBottomMargin: CGFloat = 15
+    
+    private let cityNameTextfield = UITextField()
+    private let cityNamesTableView = UITableView()
+    var cityNameArray: [String] = []
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        // Do any additional setup after loading the view.
+        setupView()
+        setupConstraints()
+        cityNameTextfield.addTarget(self, action: #selector(loadCityNames), for: .editingChanged)
+        
+    }
+    
+    @objc func loadCityNames() {
+        guard let cityName = cityNameTextfield.text else {return}
+        getCityName(fromCity: cityName, completion: {(location) -> Void in
+            guard let results = location else {return}
+            if !results.isEmpty {
+                var cityArray: [String] = []
+                for i in 0...(results.count-1){
+                    guard let city = results[i].city else {break}
+                    cityArray.append(city)
+                    self.cityNameArray = cityArray
+                }
+            } else {
+                return
+            }
+          })
+        cityNamesTableView.reloadData()
+    }
+    
+    func getCityName(fromCity cityName: String, completion: @escaping (_ result: [Location]?) -> Void) {
+        guard let queryURL =  URL(string:"http://geodb-free-service.wirefreethought.com/v1/geo/cities?&namePrefix=" + cityName + "&sort=-population") else {completion(nil); return}
+        let session = URLSession.shared
+        
+        session.dataTask(with: queryURL, completionHandler: { [self] data, response, error -> Void in
+            
+            if (error != nil) {
+                completion(nil)
+            }
+            
+            if let httpResponse = response as? HTTPURLResponse {
+                if httpResponse.statusCode == 200 {
+                    do {
+                        guard let data = data else { return }
+                        let jsonResult = try JSONSerialization.jsonObject(with: data, options: JSONSerialization.ReadingOptions.mutableContainers)
+                        self.decodeJson(jsonResult: jsonResult, completion: completion)
+                    } catch let e {
+                        print(e)
+                        completion(nil)
+                    }
+                } else {
+                    completion(nil)
+                }
+            } else {
+                completion(nil)
+            }
+            
+        }).resume()
+    }
+    
+    func decodeJson(jsonResult: Any,completion: @escaping (_ result: [Location]?) -> Void) {
```

Comment:
ta funkcja nie powinna przyjmować completion, ona powinna mieć za zadanie tylko jedną rzecz, to znaczy dekodowane jsona. Przyjmuje JSONA i zwraca zdecodowany obiekt.